### PR TITLE
WT-9915 For tiered storage testing, get predictable outputs for test/format

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@keitharnoldsmith, @sueloverso Here's the changes to test/format for predictable replay.  It looks large, but a large proportion of changes are mechanical.  I'd be happy to talk through the change, or you can read the big comment in replay.c to see if that makes sense.  I will still probably need to add my test and comparison scripts and a smoke test for this, but everything you see here is ready to review.

@sulabhM, I didn't mark you as a reviewer, but you're welcome to take a look, since you now know something about predictable replay!

**Comment by sueloverso** [comment]
@ddanderson I think you forgot to `git add replay.c` as it doesn't appear in the diff and all the testing cannot find it either.

**Comment by ddanderson** [comment]
@sueloverso !! just pushed replay.c.  That should help :-)

**Review by sueloverso — COMMENTED** [review]
I wanted to submit this so far. I haven't yet had a chance to go through replay.c and will do that separately.

File: test/format/config.sh [hunk 1]
```diff
@@ -20,12 +20,13 @@ typedef struct {
 
 #define C_BOOL 0x001u        /* Boolean (true if roll of 1-to-100 is <= CONFIG->min) */
 #define C_IGNORE 0x002u      /* Not a simple randomization, configured specially */
-#define C_STRING 0x004u      /* String (rather than integral) */
-#define C_TABLE 0x008u       /* Value is per table, not global */
-#define C_TYPE_FIX 0x010u    /* Value is only relevant to FLCS */
-#define C_TYPE_LSM 0x020u    /* Value is only relevant to LSM */
-#define C_TYPE_ROW 0x040u    /* Value is only relevant to RS */
-#define C_TYPE_VAR 0x080u    /* Value is only relevant to VLCS */
+#define C_NEUTRAL 0x004u     /* Does not affect data written to disk */
```
Comment (sueloverso) [inline test/format/config.sh:23]: I am a little worried about future developers either not knowing about this or forgetting to include it. What is the result/risk of someone not setting this when it should be set? Is there a clean way to force the developer to specify something? I.e. create two defines `C_DATA_TOUCH` and `C_DATA_AVOID` (or something) and require that one or the other be set always. That forces the developer to think about it.

File: test/format/replay.c [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "format.h"
+
+/*
+ * Predictable replay is ability to do test runs multiple times and always have predictable changes
+ * made at every time stamp. Two predictable runs with the same starting "data seed"" executed up to
```
Comment (sueloverso) [inline test/format/replay.c:33]: nit: `time stamp` should be `timestamp`.
Also: Fix the repeated double quote "data seed""

File: test/format/replay.c [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+ * (like bulk loading) when deciding on a random course, uses the global data RNG, which is seeded
+ * by the data seed. Global decisions that don't affect data, like whether to turn on verbose, or
+ * even the rate of checkpointing, use the global "extra" RNG, which is seeded by the "extra seed".
+ * Changing the extra seed may change some characteristics of how a workload is tested, but should
+ * not change any data on disk. When worker threads run, they have their own data and extra RNGs,
+ * and these are seeded by the timestamp they are working on.
+ *
+ * Before a worker thread can decide on what operation to do on which key in which table, it must
+ * obtain the next timestamp. Timestamps are doled out atomically, so no two worker thread can ever
+ * "work on" the same timestamp. The timestamp is XOR-ed with the data seed, the result is the seed
```
Comment (sueloverso) [inline test/format/replay.c:49]: I'd prefer fewer or no phrases with double-quotes. I suggest instead of "work on" you say `perform operations using...`

Look at removing them earlier for the seeds too.

File: test/format/replay.c [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+ * of the thread's private data RNG for the duration of that operation. Likewise, a private extra
+ * RNG is seeded from the timestamp and the extra seed. This ensures that all decisions about what
+ * is committed at that timestamp is predictable based on the timestamp. As you might expect, the
+ * thread's data RNG is used to decide what operation to do, which table to use, and which key
+ * within the table. Other "random" decisions, like whether to reopen a session, or whether to
```
Comment (sueloverso) [inline test/format/replay.c:54]: Remove the double quotes from "random".

File: test/format/format_config.c [hunk 5]
```diff
@@ -1266,6 +1348,12 @@ config_statistics(void)
 static void
 config_transaction(void)
 {
+    /* Predictable replay requires timestamps. */
+    if (GV(RUNS_PREDICTABLE_REPLAY)) {
+        config_single(NULL, "transaction.timestamps=on", true);
+        config_single(NULL, "transaction.implicit=0", false);
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/format_config.c:1354]: For predictable replay in test format, specifically, the timestamp drives all decisions about what to do.  Once a thread has a timestamp that it will "work on", it (more or less) seeds its thread-local random number with that timestamp.  That determines exactly what key it will insert/remove/read/update at that timestamp along with any new value.  So we really cannot do predictable replay as designed without having timestamps.

File: test/format/replay.c [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+ * Note that once a thread has started to work on an operation at a timestamp, it cannot give up on
+ * the effort. If, for example, a rollback error naturally happens, we can rollback the transaction.
+ * However, immediately getting a new timestamp would mean that we would lose the consequences of
+ * the previous timestamp, perhaps a record would not be updated in a particular way. Thus, after a
+ * rollback, a thread starts again, using the same timestamp it had before, and it seeds its RNGs
+ * again using this timestamp. This gives full predictability, even in the face of temporary
+ * failures.
+ *
+ * To avoid the possibility that two threads work on the same key at the same time, we have the
+ * concept of "lanes", and only one thread can be working in a lane at once. There are LANE_COUNT
```
Comment (sueloverso) [inline test/format/replay.c:66]: Again remove the double quotes in this one instance.

File: test/format/replay.c [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+ * lanes, where LANE_COUNT is 2^k for some k. A thread uses a data RNG to choose the top bits of a
+ * key number, but the bottom k bits of the key number are set to the bottom k bits of the timestamp
+ * being worked. Those bottom k bits also determine the "lane" we are in. Each lane has a flag that
```
Comment (sueloverso) [inline test/format/replay.c:69]: "being worked" is awkward. Maybe '...of the timestamp being used in this transaction.'?
And again, un-quote "lane" here.

File: test/format/replay.c [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+ * Having some operations lag behind is a natural part of processing. This leads to a stable
+ * timestamp that may lag significantly. Due to the possibility of dependencies between operations,
+ * the more lag, the more chance that a rollback error occurs. Without predictable replay, this is
+ * not a problem, any operation that produces a rollback can be freely abandoned, and threads
+ * generally continue moving quickly ahead with more work. However, with predictable replay, no
+ * operation can be abandoned, and an operation that failed because of a dependency will repeatedly
+ * fail until the stable timestamp advances. For that reason, we keep calculating and moving the
+ * stable timestamp ahead at a much faster pace when predictable replay is configured. We also use
+ * an algorithm that only uses lanes that are in use to calculate the stable timestamp. This is safe
+ * and "more responsive" than the default calculation. And when there is a rollback error, we try to
```
Comment (sueloverso) [inline test/format/replay.c:90]: Again remove the double quotes.

File: test/format/replay.c [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+        g.replay_cached_committed = ts;
+        testutil_check(pthread_rwlock_unlock(&g.lane_lock));
+    }
+    return (ts);
+}
+
+/*
+ * replay_pick_timestamp --
+ *     Pick the next timestamp for this operation. That timestamp is used for any commits and also
+ *     determines which 'lane' we are in, to prevent races from occurring on operations on a single
```
Comment (sueloverso) [inline test/format/replay.c:131]: Remove the single quotes.

File: test/format/replay.c [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+ */
+static void
+replay_pick_timestamp(TINFO *tinfo)
+{
+    uint64_t replay_seed, stop_ts, ts;
+    uint32_t lane;
+    bool in_use;
+
+    /*
+     * Choose a unique timestamp for commits. When we do predictable replay. If we're here with the
```
Comment (sueloverso) [inline test/format/replay.c:143]: 'When we do predictable replay.' is not a full sentence.

**Review by sueloverso — APPROVED** [review]
I have a lot of clean up type comments. Otherwise LGTM. If there are major changes I'll rereview. Otherwise I'm good after the cleanup.

File: test/format/format.h [hunk 11]
```diff
@@ -224,14 +247,21 @@ typedef struct {
...
     uint64_t oldest_timestamp; /* Last timestamp used for oldest */
     uint64_t stable_timestamp; /* Last timestamp used for stable */
 
     uint64_t truncate_cnt; /* truncation operation counter */
 
+    uint64_t replay_start_timestamp;     /* Timestamp at the beginning of a run */
+    uint64_t timestamp_copy;             /* A copy of the timestamp, for safety checks */
+    uint64_t stop_timestamp;             /* If non-zero, stop when stable reaches this */
+    uint64_t replay_cached_committed;    /* Our committed timestamp, cached */
+    uint32_t replay_calculate_committed; /* Times before recalculating */
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/format.h:263]: fixed

File: test/format/format.h [hunk 12]
```diff
@@ -348,6 +389,7 @@ typedef struct {
     bool repeatable_reads; /* if read ops repeatable */
     bool repeatable_wrap;  /* if circular buffer wrapped */
     uint64_t opid;         /* Operation ID */
+    uint64_t replay_ts;    /* allocated timestamp for predictable replay */
     uint64_t read_ts;      /* read timestamp */
     uint64_t commit_ts;    /* commit timestamp */
     uint64_t stable_ts;    /* stable timestamp */
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/format.h:395]: fixed

File: test/format/format_timestamp.c [hunk 13]
```diff
@@ -106,7 +109,23 @@ timestamp_once(WT_SESSION *session, bool allow_lag, bool final)
     if (oldest_timestamp == 0)
         return;
 
-    if (!final) {
+    if (GV(RUNS_PREDICTABLE_REPLAY)) {
+        /*
+         * For predictable replay, we need the oldest timestamp to lag when the process exits. That
+         * allows two runs that finish with stable timestamps in the same ballpark to be compared.
+         */
+        if (stable_timestamp > 10000)
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/format_timestamp.c:117]: fixed

File: test/format/format_timestamp.c [hunk 14]
```diff
@@ -152,11 +171,22 @@ timestamp(void *arg)
...
-        random_sleep(&g.rnd, 15);
-
-        timestamp_once(session, true, false);
+        if (!GV(RUNS_PREDICTABLE_REPLAY))
+            random_sleep(&g.extra_rnd, 15);
+        else {
+            if ((rng(&g.extra_rnd) & 0x1) == 1)
+                __wt_yield();
+            else
+                __wt_sleep(0, 10000);
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/format_timestamp.c:187]: fixed

File: test/format/ops.c [hunk 15]
```diff
@@ -313,10 +356,8 @@ operations(u_int ops_seconds, bool lastrun)
         testutil_check(__wt_thread_create(NULL, &random_tid, random_kv, NULL));
     if (g.transaction_timestamps_config)
         testutil_check(__wt_thread_create(NULL, &timestamp_tid, timestamp, tinfo_list));
-
     if (g.checkpoint_config == CHECKPOINT_ON)
         testutil_check(__wt_thread_create(NULL, &checkpoint_tid, checkpoint, NULL));
-
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/ops.c:319]: fixed

File: test/format/replay.c (hunk 2 above)
Comment (ddanderson) [reply test/format/replay.c:33]: fixed

File: test/format/replay.c (hunk 3 above)
Comment (ddanderson) [reply test/format/replay.c:49]: fixed.  I've removed all quotes!  I guess I use quotes when I'd normally want to use italics or bold, I'll try to reform :-)

File: test/format/replay.c (hunk 4 above)
Comment (ddanderson) [reply test/format/replay.c:54]: done.

File: test/format/replay.c (hunk 6 above)
Comment (ddanderson) [reply test/format/replay.c:66]: done.

File: test/format/replay.c [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+         * Predictable replay, as it works now, requires that we're not in transaction when we start
+         * the loop.
+         */
+        testutil_assert(!intxn);
+
+        /*
+         * We're here at the start of the loop for one of four reasons:
+         *   1) We needed to rollback the transaction, so we didn't give up our replay timestamp,
+         * and we set the 'again' flag.
+         *   2) We successfully committed the last transaction, but our lane was 'behind',
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/replay.c:214]: done.

File: test/format/replay.c [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+         * and was skipped over, so we're obligated to perform the next timestamp in our lane.
+         * In that case, we have a replay timestamp and the 'again' flag is set.
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/replay.c:216]: done.

File: test/format/replay.c [hunk 18]
```diff
@@ -0,0 +1,467 @@
...
+    fprintf(stderr, "%s: oldest=%" PRIu64 ", stable=%" PRIu64 "\n", msg, oldest_ts, stable_ts);
+}
+#endif
+
+/*
+ * replay_run_sync --
+ *     Called at beginning and end of runs to set up the lanes.
+ */
+static void
+replay_run_sync(void)
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/replay.c:258]: How about `replay_lane_reset`?

[the remark this answers is not in the document]
Comment (sueloverso) [reply test/format/replay.c:258]: That works for me!

File: test/format/ops.c [hunk 19]
```diff
@@ -878,25 +958,59 @@ ops(void *arg)
...
+             * integrity of the replay.
+             */
+            ntries = 0;
+
+        /* Number of tries only gets incremented during predictable replay. */
+        testutil_assert(ntries == 0 || (!intxn && tinfo->replay_again));
+
+        /*
+         * In predictable replay, put each operation in its own transaction. It's possible we could
+         * make multiple operations work predictably in the future.
```
Comment (keitharnoldsmith) [inline test/format/ops.c:1003]: Does it really seem possible? I.e., is it reasonable to create a ticket so maybe we get to it in the future?

File: test/format/ops.c [hunk 20]
```diff
@@ -622,6 +688,20 @@ table_op(TINFO *tinfo, bool intxn, iso_level_t iso_level, thread_op op)
     /* Acquire a cursor. */
     tinfo->cursor = table_cursor(tinfo, table->id);
 
+    /*
+     * Predictable replay has some restrictions. Someday we may be able to resolve some of these
+     * restrictions, this may require adding complexity.
```
Comment (keitharnoldsmith) [inline test/format/ops.c:693]: Is this feasible enough to ticket as potential follow-on work. Or does it seem too intractable?

**Review by keitharnoldsmith — APPROVED** [review]
LGTM with some minor editorial comments

Overall, it looks very nice. Reading the design and chatting about this at meetings, I was expecting something harder to get my head around, but you've implemented/organized it very clearly. Nice work!

File: test/format/config.sh (hunk 1 above)
Comment (ddanderson) [reply test/format/config.sh]: As mentioned below, I removed the premise of having a "predictable" configuration be generated.  So `C_NEUTRAL` is gone!

**Comment by ddanderson** [comment]
Oops - @keitharnoldsmith, I haven't addressed your review comments - mostly comment fixups I think.  I will do that now.

File: test/format/replay.c [hunk 21]
```diff
@@ -0,0 +1,467 @@
...
+ * might expect, the thread's data RNG is used to decide what operation to do, which table to use,
+ * and which key within the table. Other random decisions, like whether to reopen a session, or
+ * whether to repeat a read from the snap list, use the extra RNG.
+ *
+ * Note that once a thread has started to work on an operation at a timestamp, it cannot give up on
+ * the effort. If, for example, a rollback error naturally happens, we can rollback the transaction.
+ * However, immediately getting a new timestamp would mean that we would lose the consequences of
+ * the previous timestamp, perhaps a record would not be updated in a particular way. Thus, after a
+ * rollback, a thread starts again, using the same timestamp it had before, and it seeds its RNGs
+ * again using this timestamp. This gives full predictability, even in the face of temporary
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/replay.c:62]: If a thread needs to do predictable operations within a transaction (and do the same operations another time, after a ROLLBACK), it will need to seed its RNG with the same value every time.  That value is based on the timestamp the thread is working on combined with the global seed.  At the time of the ROLLBACK, since it may have already used the RNG, it needs to reset it with the original seed it used.  Fortunately calculating the seed and resetting the RNG is a tiny operation.

Parts of this are covered in the previous paragraph.  Is there a better way to present it?

File: test/format/ops.c (hunk 19 above)
Comment (ddanderson) [reply test/format/ops.c:1014]: Yes, it seems possible - I'll open a ticket for discussion.

Comment (ddanderson) [reply test/format/ops.c:1014]: I put this into WT-10571 with some thoughts, it doesn't seem too hard.  It's in the followon Epic.

File: test/format/ops.c (hunk 20 above)
Comment (ddanderson) [reply test/format/ops.c:704]: This one is tougher.  When we insert a record to a column store, we expand the key space by one, and it's hard to expose that immediately to all threads in a predictable way.  We could "hide" the real end point of the column store for, say 1000 timestamps.  But we have to handle the case when the creator is behind the front runner by > 1000 timestamps.  I think a solution is to only allow the creator of a new key to have a matched lane for the new key.  I'm convincing myself that it could be possible, I'll write up a ticket.

Comment (ddanderson) [reply test/format/ops.c:704]: WT-10572.  Feels possible, but maybe too complex to pursue unless we're really motivated to more fully test CS with predictable replay.

File: test/format/replay.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (keitharnoldsmith) [reply test/format/replay.c:62]: I think the disconnect for me was that I assumed that selecting an operation meant getting a couple random numbers, using those to define the op, then executing the op. If I'm connecting the dots correctly, I think you're saying that as format performs the operation it may consult the RNG more times. 

I think the wording here is OK. Just clarifying my picture of what is happening.

**Review by keitharnoldsmith — APPROVED** [review]
Thanks Don!

LGTM

File: test/format/config.sh [hunk 22]
```diff
@@ -246,6 +250,8 @@ CONFIG configuration_list[] = {
 
 {"runs.rows", "number of rows", C_TABLE, 10, M(1), M(100)}
 
+{"runs.predictable_replay", "configure predictable replay", C_BOOL, 0, 0, 0}
```
Comment (sueloverso) [inline test/format/config.sh:253]: Alphabetize in the list please

File: test/format/format.h [hunk 23]
```diff
@@ -102,6 +102,12 @@ typedef struct {
     const char *track; /* Tag for tracking operation progress */
 } SAP;
 
+/* Arguments to the read scanner. */
+typedef struct {
+    WT_CONNECTION *conn;
+    WT_RAND_STATE *rnd;
+} READ_SCAN_ARGS;
```
Comment (sueloverso) [inline test/format/format.h:109]: Alphabetize above RWLOCK in list of random format structures.

File: test/format/format.h [hunk 24]
```diff
@@ -121,6 +127,23 @@ typedef struct {
...
+ *
+ * A more complete description of how this fits into predictable replay is in replay.c .
+ */
+typedef struct {
+    uint64_t last_commit_ts;
+    bool in_use;
+} LANE;
+#define LANE_NONE UINT32_MAX /* A lane number guaranteed to be illegal */
+#define LANE_COUNT 1024u
+
```
Comment (sueloverso) [inline test/format/format.h:146]: Alphabetize in list of structures. Also please move CONFIG if it can be moved to alphabetize.

File: test/format/format.h [hunk 25]
```diff
@@ -224,14 +247,21 @@ typedef struct {
...
+    WT_RAND_STATE extra_rnd; /* Global RNG state for extra operations */
 
     uint64_t timestamp;        /* Counter for timestamps */
     uint64_t oldest_timestamp; /* Last timestamp used for oldest */
     uint64_t stable_timestamp; /* Last timestamp used for stable */
 
     uint64_t truncate_cnt; /* truncation operation counter */
 
+    uint32_t replay_calculate_committed; /* Times before recalculating cached committed */
+    uint64_t replay_cached_committed;    /* Our committed timestamp, cached */
```
Comment (sueloverso) [inline test/format/format.h:260]: Alphabetize these two.

File: test/format/format.h [hunk 26]
```diff
@@ -311,7 +347,12 @@ typedef struct {
 
     SAP sap; /* Thread's session event handler information */
 
-    WT_RAND_STATE rnd; /* thread RNG state */
+    WT_RAND_STATE data_rnd;  /* thread RNG state for data operations */
+    WT_RAND_STATE extra_rnd; /* thread RNG state for extra operations */
+
+    thread_op op;      /* Operation */
+    uint32_t lane;     /* Current lane for replay */
```
Comment (sueloverso) [inline test/format/format.h:354]: Alphabetize these two if this group of 3 are related.

File: test/format/format_config.c [hunk 27]
```diff
@@ -335,6 +395,31 @@ config_table(TABLE *table, void *arg)
...
+                WARN("turning off truncate for table%" PRIu32 " to work with predictable replay",
+                  table->id);
+            config_single(table, "ops.truncate=0", false);
+        }
+        if (!replay_operation_enabled(MODIFY)) {
+            if (config_explicit(table, "ops.pct.modify") && TV(OPS_PCT_MODIFY))
+                WARN("turning off modify operations for table%" PRIu32
+                     " to work with predictable replay",
+                  table->id);
+            config_single(table, "ops.pct.modify=0", false);
```
Comment (sueloverso) [inline test/format/format_config.c:419]: Alphabetize these two conditionals

File: test/format/format_config.c [hunk 28]
```diff
@@ -1266,6 +1393,12 @@ config_statistics(void)
 static void
 config_transaction(void)
 {
+    /* Predictable replay requires timestamps. */
+    if (GV(RUNS_PREDICTABLE_REPLAY)) {
+        config_single(NULL, "transaction.timestamps=on", true);
+        config_single(NULL, "transaction.implicit=0", false);
```
Comment (sueloverso) [inline test/format/format_config.c:1399]: Alphabetize these two lines.

File: test/format/format_config_def.c [hunk 29]
```diff
@@ -244,6 +250,9 @@ CONFIG configuration_list[] = {{"assert.read_timestamp", "assert read_timestamp"
 
   {"runs.rows", "number of rows", C_TABLE, 10, M(1), M(100), V_TABLE_RUNS_ROWS},
 
+  {"runs.predictable_replay", "configure predictable replay", C_BOOL, 0, 0, 0,
+    V_GLOBAL_RUNS_PREDICTABLE_REPLAY},
```
Comment (sueloverso) [inline test/format/format_config_def.c:254]: Alphabetize this above. (Although I think this is generated from the other file.)

File: test/format/ops.c [hunk 30]
```diff
@@ -226,14 +246,22 @@ rollback_to_stable(WT_SESSION *session)
...
+        g.timestamp = g.stable_timestamp;
 }
 
 /*
  * operations --
  *     Perform a number of operations in a set of threads.
  */
 void
-operations(u_int ops_seconds, bool lastrun)
+operations(u_int ops_seconds, u_int run_number, u_int run_count)
```
Comment (sueloverso) [inline test/format/ops.c:264]: `run_number` and `run_count` are pretty similarly named to each other. Could you find one or two different names that makes them more distinctive.

File: test/format/ops.c [hunk 31]
```diff
@@ -291,8 +333,11 @@ operations(u_int ops_seconds, bool lastrun)
 
     /* Initialize and start the worker threads. */
     tinfo_init();
+    lanes_init();
```
Comment (sueloverso) [inline test/format/ops.c:336]: If possible, alphabetize these two init calls.

File: test/format/replay.c [hunk 32]
```diff
@@ -0,0 +1,543 @@
...
+     * threads to get into a holding state, waiting for the truncate to start and then
+     * complete before continuing with their next operation. While we could fudge this
+     * in certain ways (e.g. operations with 10000 timestamps of a truncate would be
+     * forced to stay out of its table), there still would be a lot of details, and
+     * some rethink of our lane strategy. Even getting this to work, we would have
+     * a truncate that had the whole table to itself, which doesn't seem like an
+     * effective test.
+     */
+    if (op == TRUNCATE)
+        return (false);
```
Comment (sueloverso) [inline test/format/replay.c:176]: Alphabetize TRUNCATE and its comment below REMOVE

File: test/format/replay.c (hunk 9 above)
Comment (sueloverso) [reply test/format/replay.c:218]: The quotes are still here.

File: test/format/replay.c (hunk 10 above)
Comment (sueloverso) [reply test/format/replay.c:230]: This is still here too.

File: test/format/replay.c [hunk 33]
```diff
@@ -0,0 +1,543 @@
...
+static void
+replay_pick_timestamp(TINFO *tinfo)
+{
+    uint64_t replay_seed, stop_ts, ts;
+    uint32_t lane;
+    bool in_use;
+
+    /*
+     * Choose a unique timestamp for commits. When we do predictable replay. If we're here with the
+     * 'again' flag, we already have a timestamp picked for us.
```
Comment (sueloverso) [inline test/format/replay.c:231]: To get rid of the quotes how about rewording like:
If the field for replaying again is set, we already have ...

File: test/format/replay.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (sueloverso) [reply test/format/replay.c:301]: Single quotes are still here. Can you reword and remove them from 'behind'?

File: test/format/replay.c [hunk 34]
```diff
@@ -0,0 +1,543 @@
...
+{
+    uint64_t prepare_ts, ts;
+
+    testutil_assert(GV(RUNS_PREDICTABLE_REPLAY));
+
+    prepare_ts = tinfo->replay_ts;
+    if (prepare_ts != 0) {
+        /*
+         * Except when starting a run, the read timestamp cannot advance beyond our lane's last
+         * commit timestamp, which will always be tinfo->replay_ts - LANE_COUNT. Picking a timestamp
```
Comment (sueloverso) [inline test/format/replay.c:409]: It would be good to remove this calculation from the comment. The field name and expression do not appear in the code below and it will be hard to keep accurate. Also remove the "(replay_ts)" from line 410 since it also refers to a field that is not used in the code below.

**Review by sueloverso — COMMENTED** [review]
I only have minor change requests but a bunch of them. I'd like to look at them once they're addressed.

**Comment by ddanderson** [comment]
@sueloverso I think all these issues have been addressed.  The build fails due to a known platform issue.  I have reworked the code and comment around what used to be the `tinfo->replay_ts - LANE_COUNT` comment.  Hopefully the result is clearer - let me know what you think.

File: test/format/format_config.c (hunk 27 above)
Comment (sueloverso) [reply test/format/format_config.c:419]: This one is not changed. Is there a reason we have to check truncate first or could they be alphabetized to check MODIFY and then TRUNCATE?

File: test/format/format_config.c (hunk 28 above)
Comment (sueloverso) [reply test/format/format_config.c:1399]: This is still not alphabetized. Is there a reason requiring this order?

File: test/format/format_config_def.c (hunk 29 above)
Comment (sueloverso) [reply test/format/format_config_def.c:254]: This one was missed too. I thought it was generated from config.sh but maybe not. Please fix the order.

File: test/format/ops.c (hunk 30 above)
Comment (sueloverso) [reply test/format/ops.c:264]: Did you think about this? To me `count` and `number` have a lot of potentially overlapping meaning.

**Review by sueloverso — APPROVED** [review]
There are a number of older comments that are not yet fixed. I resolved everything that is fixed. I re-iterated on some comments, but not all. But all of my outstanding comments appear to still be valid. They're all minor/nits so I'll approve and don't need to re-review.

File: test/format/config.sh (hunk 22 above)
Comment (ddanderson) [reply test/format/config.sh:251]: This is fixed, I don't know why it doesn't say Outdated.

File: test/format/format.h (hunk 23 above)
Comment (ddanderson) [reply test/format/format.h]: This is done.

File: test/format/format.h (hunk 24 above)
Comment (ddanderson) [reply test/format/format.h]: LANE and CONFIG have been moved.

File: test/format/replay.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/replay.c]: Fixed.

File: test/format/replay.c (hunk 33 above)
Comment (ddanderson) [reply test/format/replay.c]: Fixed.

File: test/format/replay.c (hunk 34 above)
Comment (ddanderson) [reply test/format/replay.c]: Fixed.

File: test/format/format_config.c (hunk 27 above)
Comment (ddanderson) [reply test/format/format_config.c:413]: My apologies - I missed your comment(s) here.  I always look at the "Conversations" page in github, and yesterday's comment "Alphabetize these two conditionals" does not appear for me.

File: test/format/format_config.c (hunk 28 above)
Comment (ddanderson) [reply test/format/format_config.c:1398]: Just fixed - sorry I did not see your earlier comment.

File: test/format/format_config.c (hunk 27 above)
Comment (ddanderson) [reply test/format/format_config.c:413]: Fixed.

File: test/format/format_config_def.c (hunk 29 above)
Comment (ddanderson) [reply test/format/format_config_def.c:252]: Yes, this is generated.  I forgot to run config.sh, which CMake doesn't check or do.

Comment (ddanderson) [reply test/format/format_config_def.c:252]: Fixed, now.

File: test/format/replay.c (hunk 32 above)
Comment (ddanderson) [reply test/format/replay.c]: Fixed.

File: test/format/replay.c [hunk 35]
```diff
@@ -0,0 +1,467 @@
...
+ * again using this timestamp. This gives full predictability, even in the face of temporary
+ * failures.
+ *
+ * To avoid the possibility that two threads work on the same key at the same time, we have the
+ * concept of "lanes", and only one thread can be working in a lane at once. There are LANE_COUNT
+ * lanes, where LANE_COUNT is 2^k for some k. A thread uses a data RNG to choose the top bits of a
+ * key number, but the bottom k bits of the key number are set to the bottom k bits of the timestamp
+ * being worked. Those bottom k bits also determine the "lane" we are in. Each lane has a flag that
+ * determines whether the lane is in use by some operation. If thread T1 working an operation at
+ * timestamp X takes a sufficiently long time relative to other operations, it may be that current
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/format/replay.c]: Fixed.

File: test/format/ops.c (hunk 31 above)
Comment (ddanderson) [reply test/format/ops.c:335]: Fixed.
